src/meta-srv/src/region/lease_keeper/mito.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use common_meta::rpc::router::{
+    convert_to_region_leader_map, convert_to_region_peers_map, RegionRoute,
+};
+use store_api::region_engine::RegionRole;
+use store_api::storage::RegionId;
+
+use crate::region::lease_keeper::utils::{inactive_follower_regions, inactive_leader_regions};
+
+/// Retains active mito regions(`datanode_regions`), returns inactive regions.
+///
+/// It removes a leader region if the `node_id` isn't the corresponding leader peer in `region_routes`.
+///
+/// It removes a follower region if the `node_id` isn't one of the peers in `region_routes`.
+pub fn retain_active_regions(
+    node_id: u64,
+    datanode_regions: &mut Vec<(RegionId, RegionRole)>,
+    region_routes: &[RegionRoute],
+) -> HashSet<RegionId> {
+    let region_leader_map = convert_to_region_leader_map(region_routes);
+    let region_peers_map = convert_to_region_peers_map(region_routes);
+
+    let inactive_region_ids = datanode_regions
+        .clone()
+        .into_iter()
+        .filter_map(|(region_id, role)| match role {
+            RegionRole::Follower => {
+                inactive_follower_regions(node_id, region_id, &region_peers_map)
+            }
+            RegionRole::Leader => inactive_leader_regions(node_id, region_id, &region_leader_map),
+        })
+        .collect::<HashSet<_>>();
+
+    let _ = datanode_regions
+        .extract_if(|(region_id, _)| inactive_region_ids.contains(region_id))
+        .collect::<Vec<_>>();
```

Comment:
You can use `count` to consume the iterator if you don't need the result. Save some space for collecting.